hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -158,9 +165,12 @@ protected QueryResult tryQueryUsingIndexes(Predicate predicate, Collection<Integ
         }
 
         QueryResult result = newQueryResult(partitions.size(), iterationType);
-        // if partition state version has changed in the meanwhile, this means migrations were executed and we may
+        // If partition state version has changed in the meanwhile, this means migrations were executed and we may
         // return stale data, so we should rather return null and let the query run with a full table scan
-        if (initialPartitionStateVersion != partitionService.getPartitionStateVersion()) {
+        // Also make sure there are no long migrations in flight which may have started after starting the query
+        // but not completed yet.
+        if (mapServiceContext.getService().getOwnerMigrationsInFlight() > 0
+                || initialPartitionStateVersion != partitionService.getPartitionStateVersion()) {
```

Comment:
Could we maybe have a helper method for `mapServiceContext.getService().getOwnerMigrationsInFlight() > 0` and for the full condition check with or? It's being repeated in a couple of places. For example?

`mapServiceContext.getService().getOwnerMigrationsInFlight() > 0` -> `areThereMigrationsInFlight()`

`initialPartitionStateVersion != partitionService.getPartitionStateVersion())` -> `hasThePartitionStateChanged(initialPartitionStateVersion)`

Both: `isResultNonAmbigous()`?